### ISSUE TITLE
Test form tool for story collection

### ISSUE THREAD

**Issue by xmatthewx** [MozillaFoundation/Design#512]
tl;dr - help test a 3rd party form tool, to see if it meets requirements as a story collection tool 

### Background

To inform the Foundation’s campaigns and insights, we sometimes solicit input from partners or the public. Our Youtube campaign demonstrated the value. This proposed tool aims to enable that, making the job easier for staff and more inviting for participants.

### Status

We have gathered requirements from stakeholders. Chris and I are currently reviewing specific options. We would like to test a few things for our strongest candidate against the requirements (detailed fully the project brief linked below). 

### Requirements

 - [ ] Quality embedded (can be in page, or pop up)
 - [ ] Branded Mozilla
 - [ ] Link to Privacy Policy
 - [ ] Flexible questions for different use cases
 - [ ] On/Off state – can it be turned off for new submissions? 

Note: this issue is scoped to test the form and our ability to brand and embed it.  Other aspects of this project are out of scope for this issue. 

Note: no tool will meet all the requirements perfectly. We do not have the capacity to build to perfection, so tradeoffs will be inevitable. 

### Links

https://docs.google.com/document/d/1CYRXutd7G_6l9clonn-ogkinwxeebqcIjx-_Wna1YHU/edit#

cc: @kristinashu

**Comment by kristinashu** [comment]
@xmatthewx Sabrina can work on this! Please just set up a kick-off meeting with her if/when Chris approves Type Form.

**Comment by xmatthewx** [comment]
Excellent. cc @cadecairos - see comment above. 

fyi: This is where we are tracking TypeForm review: https://github.com/MozillaFoundation/mofo-devops/issues/750

**Comment by xmatthewx** [comment]
Typeform has passed an initial review by myself and @cadecairos. Devops will set up an account and enable it for embeds on staging. Tracking here: https://github.com/MozillaFoundation/mofo-devops/issues/751

**Comment by kristinashu** [comment]
This is ready for design! @xmatthewx and @sabrinang to met and then she can start testing it out.

**Comment by cadecairos** [comment]
I've added `https://devopstypeform.typeform.com` to the staging server's content security policy directive for iframe embeds. LMK if it doesn't seem to be working.

**Comment by sabrinang** [comment]
Thanks Cade! I can access Typeform now, checked it out a bit and can confirm the typeform iframe embeds do appear on mofostaging. I've sent a calendar invite to meet with Matthew to coordinate.

**Comment by sabrinang** [comment]
Met with Matthew today to discuss context of what forms would be used for. Stuff we discussed and notes on things I'll explore:
- check out what we can do with the different plan tiers (free to premium)
- explore various embed/share options and test on responsive screens (button pop-ups, one question blog post to email)
- test various use cases and test branding options
- check out what the responses/data can sync with
- document exploration, tips and what we can do in a few screenshots for reference

**Comment by kristinashu** [comment]
The needs this surfaces could influence - [Component] Option to set height and hide scrollbar for iFrame #4115

**Comment by xmatthewx** [comment]
Thanks @sabrinang. You don't need to worry about data sync. I'm keen to share back with stakeholders soon to get approval and figure out budget piece. The process could be slow. Let's timebox this. If the group approves, we can start to play and experiment more. I think it's going to be fun and offer many new opportunities we didn't even expect. 

Deliverables
 - [ ] Check off requirements in 1st comment up top
 - [ ] Test one long form on staging (1-column, 2-column page) Grab screenshots. 
 - [ ] Test a single-question poll on staging between paragraphs. Grab screenshot. ex: Do you think AI robots will take over the world? Yes/No/Unsure 

Thanks!!

**Comment by sabrinang** [comment]
@xmatthewx I took notes and screenshots of the review in this doc: https://docs.google.com/document/d/18Lz6pFNtIFOImVk_fYEelfAGytFeaZtpW_sKyW6eYLQ/edit#

**Overview to answer the issue checkboxes:**
- embed on wagtail is framed by our iFrame component (any code snippets they have for popup or fullpage can not be inserted) but current iFrame cuts off form on mobile
- yes we can have mozilla branding through font, colour, images, bg, welcome/thank you screens and copy
- link to Privacy Policy can be included in intro form text, our iFrame component caption field or our paragraph component 
- 19 question types with various settings and logic jumps for different use cases
- closed and open states for responses and can have customized closed message
- long form style is unavailable – it'll always show one question at a time as it’s “designed to feel like human conversation”
- 1 column only with our iFrame component (multiple choice options can have different alignment of answers)
- 1 question among text by setting the background to white, hide the progress bar and only have a one question form _(format will always show one question at a time anyway)_

**Comment by xmatthewx** [comment]
Excellent. Thanks for doing such an amazing thorough job @sabrinang. Two questions: 
 - do you have any concerns?
 - what does your gut say? Is the tool a good choice? Where are you on a spectrum from "excited" to "it's fine"?

**Comment by xmatthewx** [comment]
also @sabrinang - hopefully someone, somewhere on the internet has solved the iframe issue, setting a fixed height on it, or an auto height based on screensize. If not, but and pop-out might have to be the route.

**Comment by xmatthewx** [comment]
I just noticed that they have an option for a left/right [drawer](https://help.typeform.com/hc/en-us/articles/360029249212-Embed-your-typeform) pop-over. I wonder if that would avoid the iframe issue. I believe you can still remove their branding.

**Comment by sabrinang** [comment]
@xmatthewx I think that's a type of overlay/animation for the pop-up but would still require us to embed their code to do so? We can try having our own button link out to the form on a different tab but I don't think we can do pop-ups currently with wagtail as there's no way to include typeform's embed code. If we have the premium account the branding will be removed _(you can still see powered by typeform when you click into the form or when it loads)_

Overall, I think it looks and functions well and my only concerns would be around embedding it with our current iFrame component as we can't use their embed code currently. If we use the forms separately it's good but should figure out how wagtail could handle embed codes or adjust iFrame component https://github.com/mozilla/foundation.mozilla.org/issues/4115

**Comment by xmatthewx** [comment]
Yes, it will still require the use of their embed code. And, we might need to update our component or create a new component to do that. But... the slide drawer will likely avoid the iframe scroll issue because it is a full height overlay. And as a result, it is possibly the best solution from both a UX and tech perspective. Consider discussing over video with mavis or tais and/or @cadecairos.

**Comment by sabrinang** [comment]
Ok, I'll reach out to devs regarding component options. Yeah +1 for using the full height to give the content more space to avoid scrollbars when possible.

**Comment by cadecairos** [comment]
RRA bug filed https://bugzilla.mozilla.org/show_bug.cgi?id=1630701

**Comment by cadecairos** [comment]
@sabrinang I prototyped a component for typeform based on the airtable component.

The PR is here: https://github.com/mozilla/foundation.mozilla.org/pull/4526 - you can grab test credentials for the CMS backend from our review-app slack channel 

The test page is: https://foundation-s-typeformco-al8pi5.herokuapp.com/en/campaigns/typeform-test/

**Comment by sabrinang** [comment]
Thanks for sharing the test page @cadecairos, I checked it out:
 
On firefox desktop, I see a scroll bar when the window is narrow and on hover _(it isn't there on chrome though)_:
<img width="350" alt="screenshot - narrow desktop form scrollbar" src="https://user-images.githubusercontent.com/2492510/79799299-151a1380-830f-11ea-8037-17ef80b07629.gif">

On mobile, firefox android and chrome I see "embedding prototype form" heading and start button below the viewport on my pixel 2 device:
<img width="200" alt="screenshot-embed-mobile" src="https://user-images.githubusercontent.com/2492510/79799291-12b7b980-830f-11ea-8a1a-0a616d5f09c8.jpg">

- After clicking the start button, the form is fullscreen though
<img width="200" alt="fullscreen-form" src="https://user-images.githubusercontent.com/2492510/79800412-03397000-8311-11ea-9bd2-dcc445e02e26.png">

___________________

- It would be preferable if the scrollbar didn't appear _(not sure if this will change if the form content is long/short or if it is a browser issue)_ 
- On mobile, if the form content itself or button appears without the extra space and heading, I think that would be preferable

**Comment by sabrinang** [comment]
Note: use cases for this form tool
- would of used this for feedback/poll (instead of google forms) for Zoom, YT Regrets, Venmo
- Jen would of used it for the 'what can we do right now' and potentially for upcoming love letters project

Will look into embed/iframe/responsive issues in dev issue here: https://github.com/mozilla/foundation.mozilla.org/issues/4652

**Comment by sabrinang** [comment]
@mmmavis' test page: https://foundation-s-mavis-type-atio4y.herokuapp.com/en/campaigns/typeform-test/

**Feedback:**
Desktop
- nice, scrollbar only appears on long questions past iFrame now
- Long questions: when scrolling down can we have the question sticky? (like how typeform does embeds – see below) 
![typeform-sticky](https://user-images.githubusercontent.com/2492510/82838791-214f3e80-9e82-11ea-82d0-f9641decf42a.gif)

Mobile
- nice, form first page appears now (without heading and extra space) and tapping start makes form full screen 
- same feedback, when form is full screen on a long question can we have the question sticky above?

**Comment by mmmavis** [comment]
@sabrinang The one you've tested before is the embedded **widget**. Because this is a third-party embed, Foundation site script basically has no control with anything in the iframe and can't make the iframe fully responsive. As I mentioned in Slack, whatever you saw previously is the best we can do with the widget embed(that is, having a `height` field in the CMS Typeform streamfield so users can preset the height they want the iframe to be).

I've updated the test page to use Typeform's **popup** instead. Can you let me know if it's the popup or widget you prefer?

**Comment by sabrinang** [comment]
@mmmavis 
- If we use the embed widget, I’d also lean towards using the pop-up method to avoid iFrame cut off issues too but could we have the first page of the form be present and when the user clicks “start” it launches the pop-up from there instead? (like it did on mobile to provide more context from the form itself and one less click to get into form)

- If we use the popup button widget, I think this would potentially be the most versatile within the page itself - could users write their own button copy cta? and I assume they’ll write copy on the cms page itself to give context to the form (page is independent of the form content itself).

With those ideas in mind, I assume that the user would provide more context on the page as use cases and I like the focus on the form when it is a full screen popup over cut off content so popup is what I prefer.

**Comment by mmmavis** [comment]
> If we use the embed widget, I’d also lean towards using the pop-up method to avoid iFrame cut off issues too but could we have the first page of the form be present and when the user clicks “start” it launches the pop-up from there instead? (like it did on mobile to provide more context from the form itself and one less click to get into form)

This is something Typeform script controls. The only thing we can "customize" on our end is the fixed iframe height, and according to the [API doc](https://developer.typeform.com/embed/modes/#widget-mode), whether or not we wanna hide header/footer, background opacity level, and the button text (default as `Start`).

> If we use the popup button widget, I think this would potentially be the most versatile within the page itself - could users write their own button copy cta? and I assume they’ll write copy on the cms page itself to give context to the form (page is independent of the form content itself).

Yes, it should be possible for users to write their own button copy. Do you prefer to let the users to decide the button type they want to use? Or should we default it to either the primary button type or secondary button type?

**Comment by sabrinang** [comment]
I was thinking it will be similar to current Linkbutton component where primary button styling is default but secondary is also an option:
![image](https://user-images.githubusercontent.com/2492510/82854442-47460480-9ebd-11ea-927c-d3a996e65e87.png)

**Comment by mmmavis** [comment]
@sabrinang no problem. I'll update my PR and flag you for review!
